### PR TITLE
ARROW-8213: [Python][Dataset] Opening a dataset with a local incorrect path gives confusing error message

### PR DESCRIPTION
**Pull request by kszucs**
Workaround until it is properly handled in the C++ implementation.

### CODE REVIEW

**Comment by pitrou** [comment]
Well, can you try to fix it in the C++ implementation? :-) Or do you want me to do it?

**Comment by jorisvandenbossche** [comment]
> Well, can you try to fix it in the C++ implementation? :-) Or do you want me to do it?

I thought you didn't want to deal with non-URIs on the C++ FileSystem level?

**Comment by pitrou** [comment]
There's a C++ function called `FileSystemFromUriOrPath`. I don't remember why the Python side doesn't use it.

File: python/pyarrow/dataset.py [hunk 1]
```diff
@@ -155,19 +155,16 @@ def partitioning(schema=None, field_names=None, flavor=None):
...
-        # first check if the file exists as a local (relative) file path
-        filesystem = LocalFileSystem()
         try:
-            infos = filesystem.get_file_info([path])[0]
-        except OSError:
-            return FileSystem.from_uri(path)
-
-        if infos.type == FileType.NotFound:
-            return FileSystem.from_uri(path)
+            scheme, _ = path.split('://', 1)  # noqa
```
Comment (pitrou) [inline python/pyarrow/dataset.py:162]: This is definitely wrong. URI parsing is more complex than this...

Comment (kszucs) [reply python/pyarrow/dataset.py:162]: Well, FileSystemFromUriOrPath is used by the dataset implementation, and it raises FileNotFoundError only for `file://relate-path` uris.

Comment (kszucs) [reply python/pyarrow/dataset.py:162]: Otherwise it raises here https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/uri.cc#L207

Comment (kszucs) [reply python/pyarrow/dataset.py:162]: @pitrou any suggestions how to raise FileNotFoundError properly?

Comment (pitrou) [reply python/pyarrow/dataset.py:162]: I'll take a look.

Comment (kszucs) [reply python/pyarrow/dataset.py:162]: Thanks!

**Comment by pitrou** [comment]
@jorisvandenbossche Opinions on this?

**Comment by jorisvandenbossche** [comment]
Thanks for that fix! Yeah, I think catching "empty scheme" in the error message is the most straightforward way to handle it

**Comment by kszucs** [comment]
I don't like this kind of error handling because it is error prone, bound to an error message instead of an error type or kind - I'm fine with it as a workaround though.

**Review by wesm — APPROVED** [review]
+1, this seems okay to me, we can iterate/improve this logic surely

**Pull request closed without merging**